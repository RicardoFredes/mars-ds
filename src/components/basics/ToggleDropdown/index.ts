/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 * run "yarn reindex" to update
 */

import ToggleDropdown from "./toggle-dropdown.component";
export default ToggleDropdown;
export * from "./toggle-dropdown.types";
