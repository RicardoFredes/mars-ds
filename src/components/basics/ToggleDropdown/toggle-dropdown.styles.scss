.toggle-dropdown {
  position: relative;
  --x: 0px;
  --y: 8px;
  &--is-above {
    --y: calc(-100% - 40px - 8px);
  }
  &--is-right {
    --x: calc(-100% + 40px);
  }
  &__list {
    position: absolute;
    z-index: 101;
    min-width: 200px;
    transform: translate(var(--x), var(--y));
  }
  &:not(.toggle-dropdown--is-open) {
    .toggle-dropdown {
      &__list {
        visibility: hidden;
        pointer-events: none;
      }
    }
  }
}
