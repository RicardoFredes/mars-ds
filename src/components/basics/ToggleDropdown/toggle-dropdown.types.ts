import type { DropdownMenuItemProps } from "../DropdownMenuItem";
import type { ToggleButtonProps } from "../ToggleButton";
import type { HTMLProps } from "react";

export interface ToggleDropdownProps
  extends Omit<HTMLProps<HTMLDivElement>, "list" | "children" | "defaultValue"> {
  list: DropdownMenuItemProps[];
  componentLink?: React.ElementType | string;
  toggleButton?: ToggleButtonProps;
  disabled?: boolean;
  onToggle?: (isOpen: boolean) => void;
  defaultValue?: boolean;
}
