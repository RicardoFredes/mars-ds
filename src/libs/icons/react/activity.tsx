/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;