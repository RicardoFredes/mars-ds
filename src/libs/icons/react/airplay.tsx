/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;