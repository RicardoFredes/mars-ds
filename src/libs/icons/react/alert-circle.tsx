/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAlertCircle = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <circle cx="12" cy="12" r="10" />
  <line x1="12" y1="8" x2="12" y2="12" />
  <line x1="12" y1="16" x2="12.01" y2="16" />
</svg>
);
  
export default IconAlertCircle;