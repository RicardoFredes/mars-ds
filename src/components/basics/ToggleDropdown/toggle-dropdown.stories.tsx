import type { ComponentMeta, ComponentStory } from "@storybook/react";

import { action } from "@storybook/addon-actions";

import ToggleDropdown from "./toggle-dropdown.component";

export default {
  title: "Components/ToggleDropdown",
  component: ToggleDropdown,
  argTypes: {
    className: {
      description: "The class name of the component receives the same 'classnames' props",
      control: { type: "object" },
    },
    style: {
      description: "The style of the component receives the same 'style' props",
      control: { type: "object" },
    },
    disabled: {
      description: "Disable the toggle dropdown",
      control: { type: "boolean" },
      defaultValue: false,
    },
    defaultValue: {
      description: "The default value of the toggle dropdown",
      control: { type: "boolean" },
      defaultValue: false,
    },
    toggleButton: {
      description: "The props to configure a toggle button of the toggle dropdown",
      control: { type: "object" },
      defaultValue: {
        title: "Abrir mais opções para o item",
        iconName: "more-vertical",
        size: "sm",
        variant: "text",
      },
    },
    list: {
      description: "The list of the dropdown menu",
      control: { type: "array" },
      defaultValue: undefined,
    },
    onToggle: {
      description: "The function to be called when the toggle dropdown is toggled",
      control: { type: "function" },
      defaultValue: undefined,
      action: "onToggle",
    },
    componentLink: {
      description: 'The component link of the dropdown menu items if has "href"',
      control: { type: "object" },
      defaultValue: undefined,
    },
    isRight: {
      description: "The dropdown menu is positioned to the right",
      control: { type: "boolean" },
      defaultValue: false,
    },
    isAbove: {
      description: "The dropdown menu is positioned above",
      control: { type: "boolean" },
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof ToggleDropdown>;

const Template: ComponentStory<typeof ToggleDropdown> = (args) => <ToggleDropdown {...args} />;

const DefaultArgs = {
  list: [
    { label: "Item 1", onClick: action("Item 1 clicked") },
    { label: "Item 2", onClick: action("Item 2 clicked") },
    {
      label: "Item 3",
      onClick: action("Item 3 clicked"),
      href: "https://www.mesalva.com",
      target: "_blank",
    },
  ],
};

export const Default = Template.bind({});
Default.args = DefaultArgs;

export const Disabled = Template.bind({});
Disabled.args = { ...DefaultArgs, disabled: true };

export const Opened = Template.bind({});
Opened.args = { ...DefaultArgs, defaultValue: true };

export const Right: ComponentStory<typeof ToggleDropdown> = (args) => (
  <div className="flex justify-content-center">
    <ToggleDropdown {...args} />
  </div>
);
Right.args = { ...DefaultArgs, defaultValue: true, isRight: true };

export const Above: ComponentStory<typeof ToggleDropdown> = (args) => (
  <div className="flex justify-content-center" style={{ paddingTop: 200 }}>
    <ToggleDropdown {...args} />
  </div>
);
Above.args = { ...DefaultArgs, defaultValue: true, isAbove: true };
