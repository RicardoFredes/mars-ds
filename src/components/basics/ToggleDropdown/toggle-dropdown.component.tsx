import type { ToggleDropdownProps } from "./toggle-dropdown.types";

import classNames from "classnames";
import { useState } from "react";
import ClickOut from "react-simple-clickout";

import { CardElevations } from "../Card";
import DropdownMenu from "../DropdownMenu";
import ToggleButton from "../ToggleButton";

const ToggleDropdown = ({
  className,
  componentLink,
  disabled,
  toggleButton,
  onToggle,
  list,
  defaultValue = false,
  isRight = false,
  isAbove = false,
  ...props
}: ToggleDropdownProps) => {
  const [isOpen, setIsOpen] = useState(defaultValue);
  const hasList = Array.isArray(list) && list.length > 0;

  const toggleOpen = () => {
    if (disabled) return;
    setIsOpen(!isOpen);
    onToggle?.(!isOpen);
  };

  const cn = classNames("toggle-dropdown", className, {
    "toggle-dropdown--is-open": isOpen,
    "toggle-dropdown--is-right": isRight,
    "toggle-dropdown--is-above": isAbove,
  });

  return (
    <div className={cn} {...props} onKeyDown={(e) => (e.key === "Escape" ? toggleOpen() : null)}>
      <ToggleButton
        title="Abrir mais opções para o item"
        iconName="more-vertical"
        size="sm"
        variant="text"
        {...toggleButton}
        disabled={!hasList || disabled}
        onClick={toggleOpen}
      />

      {isOpen && !disabled && (
        <ClickOut onClickOut={toggleOpen}>
          <DropdownMenu
            elevation={CardElevations.High}
            className="toggle-dropdown__list"
            data-testid="toggle-dropdown-list"
            size="sm"
            componentLink={componentLink}
            list={list}
            onClick={toggleOpen}
          />
        </ClickOut>
      )}
    </div>
  );
};

export default ToggleDropdown;
