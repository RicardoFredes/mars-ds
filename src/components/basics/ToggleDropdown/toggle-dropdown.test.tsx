import type { ToggleDropdownProps } from "./toggle-dropdown.types";

import { render } from "@testing-library/react";

import ToggleDropdown from "./toggle-dropdown.component";

const defaultProps = {
  list: [],
};

const makeSut = (props?: ToggleDropdownProps) =>
  render(<ToggleDropdown {...defaultProps} {...props} />);

describe("<ToggleDropdown>", () => {
  it("should render component", () => {
    const wrapper = makeSut();
    expect(wrapper).toMatchSnapshot();
  });
});
