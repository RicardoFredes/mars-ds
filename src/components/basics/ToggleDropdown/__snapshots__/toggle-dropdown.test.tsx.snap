// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleDropdown> should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="toggle-dropdown"
      >
        <button
          class="btn btn--neutral btn--size-sm btn--is-disabled toggle-btn toggle-btn--variant-text"
          disabled=""
          title="Abrir mais opções para o item"
          type="button"
        >
          <span
            class="icon btn__icon"
            data-testid="more-vertical"
          >
            <svg
              height="18"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                data-name="more-vertical"
              >
                <circle
                  cx="12"
                  cy="12"
                  fill="currentColor"
                  r="2"
                />
                <circle
                  cx="12"
                  cy="5"
                  fill="currentColor"
                  r="2"
                />
                <circle
                  cx="12"
                  cy="19"
                  fill="currentColor"
                  r="2"
                />
              </g>
            </svg>
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="toggle-dropdown"
    >
      <button
        class="btn btn--neutral btn--size-sm btn--is-disabled toggle-btn toggle-btn--variant-text"
        disabled=""
        title="Abrir mais opções para o item"
        type="button"
      >
        <span
          class="icon btn__icon"
          data-testid="more-vertical"
        >
          <svg
            height="18"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              data-name="more-vertical"
            >
              <circle
                cx="12"
                cy="12"
                fill="currentColor"
                r="2"
              />
              <circle
                cx="12"
                cy="5"
                fill="currentColor"
                r="2"
              />
              <circle
                cx="12"
                cy="19"
                fill="currentColor"
                r="2"
              />
            </g>
          </svg>
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
